process_queue: Improve handling of exceptions in process_queue.

mateuszmandera opened the pull request:
https://chat.zulip.org/#narrow/stream/9-issues/topic/.E2.9C.94.20'Last.20active'.20not.20worked/near/1275878

Unhandled exceptions propagating to process_queue were not caught there,
causing improper logging - errors didn't land in errors.log as expected.
Exceptions should be caught and explicitly logged by the process_queue
logger.

Furthermore, in multi-threaded mode we want the autoreload mechanism to
be working - which it doesn't without catching the exceptions. The
correct approach is to - again - catch the exception, log it and then
send SIGUSR1 signal to trigger exit and autoreload.

Tested by modifying a worker to throw an exception at the start and then running `process_queue --queue_name X` (single-process mode) and `process_queue --multi_threaded X` (thread mode) in dev environment and verifying that the traceback gets logged by the `process_queue` logger and lands in `errors.log` as expected.

timabbott commented:
This lgtm; but I'd like a review from Alex as well.

alexmv commented:
This looks good in substance -- thanks @mateuszmandera!  I take it that it doesn't work to put `with log_and_exit_if_exception` around the `Threaded_worker(queue_name).start()` because that's in the wrong thread?  Having it twice in `Threaded_worker.__init__` and `.run` made me wonder if we can move it up a scope, essentially.

Oh, and I'd very much +1 something which renamed `class Threaded_worker` into something slightly more pythonish. :)

I think it's worth mentioning in the commit message that this is specifically about how we handle exceptions during queue worker setup, not for exceptions when consuming.  The specific case that caused us to make this PR was failure to connect when setting up the `email_senders` worker, which is currently the only worker that does work in `__init__`, so probably worth calling out in the commit message somewhere, to make the problem that this is fixing more concrete.

mateuszmandera commented:
@alexmv 

>  take it that it doesn't work to put `with log_and_exit_if_exception` around the `Threaded_worker(queue_name).start()` because that's in the wrong thread? Having it twice in `Threaded_worker.__init__` and `.run` made me wonder if we can move it up a scope, essentially.

Yeah, that happens in the wrong thread, so it wouldn't catch exceptions that would happen in `setup()`. 

Renamed the worker to, not very creatively, `ThreadedWorker` :smile: And added a mention of this being specifically about the bug of exceptions not getting caught during worker startup to the commit message.

alexmv commented:
Thanks for those tweaks, @mateuszmandera!  Merged.
